0000-conditional_dependencies.md

```diff
@@ -0,0 +1,142 @@
+
+# Detailed design
+[design]: #detailed-design
+
+## Conditional dependencies in `Cargo.toml`
+
+When declaring a dependency in a Cargo.toml file, if that dependency is an
+object, it can contain the `conditional` member, which is a boolean, similar to
+the `optional` dependency. For example:
+
+```toml
+[dependencies]
+foobar = { version = "1.0.0", conditional = true }
+```
+
+Or:
+
+```toml
+[dependencies.foobar]
+version = "1.0.0"
+conditional = true
+```
+
+A single dependency objecy cannot contain both a `conditional` key and an
+`optional` key.
+
+When generating the dependency graph, cargo will not include dependencies
+tagged conditional. It will then traverse the graph looking for conditional
+dependencies; if a matching source for a conditional dependency (e.g. a
+matching version number) is already present in the graph, cargo will
+automatically add that dependency.
+
+## `#[cfg(dependency=)]`
+
+A new kind of cfg attribute will be added to rustc. The `dependency` attribute
+will only be compiled when a certain dependency has been explicitly passed to
+rustc using the --extern flag.
+
+Because cargo automatically passes dependencies explicitly with this command,
+this code will be compiled only when cargo triggers the conditional dependency.
+
+## Convention for using conditional dependencies
+
+The above to features can be composed in any way, but to assist in adopting
+this functionality, we can define an encouraged convention.
+
+If you want to add a new conditional dependency, you should add a top level
+submodule (that is, mounted directly under lib, main, etc) with the same name
+as the conditional dependency. This module will be tagged with the cfg, and the
+extern crate declaration and all code that uses symbols from that crate will
+go inside of that module.
+
+For example:
+
+```rust
+/// lib.rs
+
+#[cfg(dependency=foobar]
+mod foobar;
+```
+
+```rust
+/// foobar.rs
+
+extern crate foobar;
+
+impl foobar::Bar for MyType {
+    ...
+}
+```
+
+## Preventing cyclic dependencies on upload to crates.io
+
+When packaging and uploading a crate to crates.io, some strategy will be used
+to ensure that its conditional dependencies could not introduce a cycle into
+the dependency graph of your crate.
+
+## `cargo doc` compiles with all dependencies
+
+Unlike other forms of compilation, the `cargo doc` command will treat
+conditional dependencies as present by default, in order to document the APIs
+which exist only when conditional dependencies are present. The conditional
+dependency may be provided somehow.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds more functionality to both cargo and rustc. The distinction between
+"conditional" dependencies and "optional" dependencies is sort of subtle and
+will have to be explained to users. More complexity always has a downside.
+
+# Alternatives
+[alternatives]: #alternatives
+
+We considered options like allowing the orphan rules to be broken by certain
```

Comment:
> I'm open to any way to resolve the papercuts of the orphan rules that doesn't break that rule. 

Yeah, I think we can design this to work without cargo. But this is pretty off topic, just wanted to point out (like sgrif) that the orphan rules have a few more problems that this can't help with.